### ISSUE TITLE
Support markdown front-matter, or ignore "bad" sub-documents

### ISSUE THREAD

**Issue by mralusw** [mikefarah/yq#794]
**Please describe your feature request.**

Markdown files (e.g. as supported by jekyll or pandoc) can have a front-matter section, which looks like a first YAML document (delimited by `---` on both ends). yq-4.7.1 bombs out on such files, because the "second" embedded document (after the `---` closing-tag of the front-matter) is not actually valid YAML (most likely problem: colons). I can't find a way to make `yq` ignore everything but the first document (e.g. `yq eval 'select(di == 0)'` still errors out).

I'd like to use `yq` to process such documents (e.g. alter jekyll post metadata, pandoc options etc).

**Describe the solution you'd like**

If we have data1.yml like:
(please keep to around 10 lines )

```yaml
---
title: 'Some post'
date: 2020-07-07
---

1st line
A colon: will break yq
```
there should be a way to call yq (e.g. `yq-4.7.1 eval 'select(di == 0)' /tmp/x.md`) without getting `Error: yaml: line 7: mapping values are not allowed in this context`. Maybe `--continue-on-errors`, `--ignore-extra-documents`, `--ignore-bad-docs`: whatever.

Note that in the example above, removing the `1st line` unbreaks yq.

If there already is a solution, please make the documentation more prominent -- I couldn't find it

**Comment by mikefarah** [comment]
`yq` uses v3 https://github.com/go-yaml to parse yaml, and it's underlying Decode has a hissy fit when it attempts to decode a bad document in a yaml file. I tried ignoring the error and parsing the next doc but it won't go.

I found a related issue raised in go-yaml https://github.com/go-yaml/yaml/pull/380 but it hasn't been touched since 2018.

**Comment by mikefarah** [comment]
Actually reading that PR it looks like its for v2 of yaml not v3.

**Comment by mralusw** [comment]
I haven't look at the go source. But from what you describe, ignoring the error and dealing with the *already parsed* (not subsequent) documents would suffice to be able to interact with the frontmatter (which is document 0).

Alternatively I should be able to restrict parsing to just documents `N1-N2,N3` etc (just 0 in this case) via some `--flag`.

**Comment by mikefarah** [comment]
Yeah I could change it to ignore the error and just deal with what's already parsed.

Only issue with the is if the front matter came before the yaml content then you couldn't parse it. But I'm guessing that's not the case for jekyll?

I could add a flag '--ignore-bad-docs' or something I guess

**Comment by mralusw** [comment]
The front-matter comes at the top and is the only YAML content in a markdown file (again, not just for jekyll &mdash; pandoc and other processors also use it, it's something of a standard). A "front-matter-prefaced" markdown file consists of
- `---`
- YAML
- `---`
- markdown content which is *not* YAML and confuses YAML parsers

See my example in the initial report. So it looks like one valid YAML document + another invalid one (which is the real markdown).

**Comment by mikefarah** [comment]
Thinking about this a little more - there is one problem. Assuming we had this flag and `yq` stopped processing the file after reading the first doc, then it will only output the first doc. Updating the yaml will work - but `yq` will not output non yaml content.

Using `--inplace` will replace the file with just the yaml content. :(

**Comment by mikefarah** [comment]
I think it'd be better to pre-process the file to split the content, use `yq` to make the changes needed and then stick the content back together.

**Comment by gkthiruvathukal** [comment]
Just want to upvote this issue and suggest a nice general approach to address it in the interim without modifying `yq`.

The YAML frontmatter is usually associated with Markdown documents--but not limited to them. Ideally, you just apply `yq` to the YAML block after extracting it.  I'm using the nicely done `python-frontmatter` and `PyYAML` projects, which can both be installed via `pip` from PyPI. This project is able to handle all of the ways of representing frontmatter, e.g. `---` and `...`, etc.

Here is a gist I created for this with a script `extract-yaml.py`.

https://gist.github.com/gkthiruvathukal/cd721d80041665d053ac6350124e137b

Here's how I used it to do a `yq` on one of my Markdown files:

```
cat 050-Paleolithic-General.md| python extract-yaml.py | yq eval .tags -
```
